### PR TITLE
Add qt6-multimedia package

### PR DESCRIPTION
**Pull request by mmcauliffe**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:110]:
```suggestion
    - libxcb                             # [linux]
    - xorg-libx11                        # [linux]
```

**Comment by mmcauliffe** [comment]
Ok, added your changes!

For Mac it's erroring on

```
$PREFIX/include/libavcodec/avcodec.h:2142:10: error: typedef 'AVMediaType' cannot be referenced with a enum specifier
```

Is the section below from qt-main required? Or is this a result from it not using the conda_build_config values?

```
    # Ensure that build and host have the same clang version on osx
    # otherwise qt6 may be out of sync compared to the pinned compiler
    # version
    - libclang     {{ cxx_compiler_version }}     # [osx]
    - libclang-cpp {{ cxx_compiler_version }}     # [osx]
    - libclang                                    # [not osx]
    - libclang-cpp                                # [not osx]
```

For  Windows, it's erroring on lines like:

```
ERROR (qt6-multimedia,Library/lib/qt6/plugins/multimedia/ffmpegmediaplugin.dll): $RPATH/api-ms-win-core-libraryloader-l1-2-1.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
```

Any ideas?

File: recipes/qt-multimedia/build.sh
Comment (hmaarrfk) [inline recipes/qt-multimedia/build.sh:21]:
```suggestion
  -DCMAKE_FIND_FRAMEWORK=NEVER \
```
try maybe this?

Comment (hmaarrfk) [reply recipes/qt-multimedia/build.sh:21]: for OSX that is.

Comment (hmaarrfk) [reply recipes/qt-multimedia/build.sh:21]: it might be taht their CMAKE file overrides this somehow...

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:108]:
```suggestion
    - xorg-libx11                        # [linux]
    - xorg-libxrandr                     # [linux]
```

**Comment by hmaarrfk** [comment]
The windows warning appears in the qt6-main changes you are proposing too, but we didn't enable "strictness" on this front for "old recipes".

Thus it is being "silently dropped as a warning" and not "stopping building as an error".

```
2024-04-10T04:23:29.3853276Z WARNING (qt6-main,Library/lib/qt6/plugins/multimedia/ffmpegmediaplugin.dll): $RPATH/api-ms-win-core-libraryloader-l1-2-0.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
2024-04-10T04:23:29.3855925Z .. is this binary repackaging?
2024-04-10T04:23:29.4590237Z WARNING (qt6-main,Library/lib/qt6/plugins/multimedia/ffmpegmediaplugin.dll): $RPATH/api-ms-win-ntuser-sysparams-l1-1-0.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
```

For osx: Its strange to me that the error isn't getting triggered in the qt-main for the main build:
https://forums.developer.apple.com/forums/thread/83067

Other posts suggest editing the files. maybe homebrew has a patch?

**Comment by hmaarrfk** [comment]
for windows, i'm nto too worried, we can raise an issue if this is the blocker, and maybe decide to move forward without it.

**Comment by mmcauliffe** [comment]
I don't have a mac, but I might be able to set up a VM this evening for it.
Is it possible that it needs to be rerendered in its own feedstock before
Mac and Windows will work? Is it worth just releasing the Linux version
first and trying to get Mac and Windows added later? The cross compiled
builds will have to be added after the fact anyway.

**Comment by mmcauliffe** [comment]
@hmaarrfk I got the OSX build working locally by removing `-DCMAKE_UNITY_BUILD=ON -DCMAKE_UNITY_BUILD_BATCH_SIZE=32`, I'm not sure what impact this has overall, but it was trying to compile some unity files in QFFmpegPlugin with objcxx:

```
2024-04-10T20:07:59.3806040Z [156/210] Building OBJCXX object src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o
2024-04-10T20:07:59.3909980Z FAILED: src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o 
2024-04-10T20:07:59.4018900Z $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -DGL_SILENCE_DEPRECATION -DQFFmpegMediaPlugin_EXPORTS -DQT_COMPILING_FFMPEG -DQT_CORE_LIB -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_UP_TO=0x050000 -DQT_EXPLICIT_QFILE_CONSTRUCTION_FROM_PATH -DQT_GUI_LIB -DQT_MULTIMEDIA_LIB -DQT_NETWORK_LIB -DQT_NO_AS_CONST -DQT_NO_AS_CONST=1 -DQT_NO_CONTEXTLESS_CONNECT=1 -DQT_NO_DEBUG -DQT_NO_EXCEPTIONS -DQT_NO_FOREACH -DQT_NO_FOREACH=1 -DQT_NO_JAVA_STYLE_ITERATORS -DQT_NO_NARROWING_CONVERSIONS_IN_CONNECT -DQT_NO_QEXCHANGE -DQT_PLUGIN -DQT_USE_QSTRINGBUILDER -DQT_WARN_DEPRECATED_UP_TO=0x070000 -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I$SRC_DIR/build/src/plugins/multimedia/ffmpeg/QFFmpegMediaPlugin_autogen/include -I$SRC_DIR/src/plugins/multimedia/ffmpeg -I$SRC_DIR/build/src/plugins/multimedia/ffmpeg -I$SRC_DIR/build/include -I$SRC_DIR/src/plugins/multimedia/ffmpeg/../darwin -I$SRC_DIR/src/plugins/multimedia/ffmpeg/../darwin/camera -I$SRC_DIR/build/src/multimedia -I$SRC_DIR/build/include/QtMultimedia/6.7.0 -I$SRC_DIR/build/include/QtMultimedia/6.7.0/QtMultimedia -I$SRC_DIR/build/include/QtMultimedia -isystem $PREFIX/include/qt6/QtCore/6.7.0 -isystem $PREFIX/include/qt6/QtCore/6.7.0/QtCore -isystem $PREFIX/include/qt6/QtCore -isystem $PREFIX/include/qt6 -isystem $PREFIX/lib/qt6/mkspecs/macx-clang -isystem $PREFIX/include/qt6/QtGui/6.7.0 -isystem $PREFIX/include/qt6/QtGui/6.7.0/QtGui -isystem $PREFIX/include/qt6/QtGui -isystem $PREFIX/include/qt6/QtNetwork -isystem $PREFIX/include -x objective-c++ -DNDEBUG -O2 -std=c++17 -isysroot /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -mmacosx-version-min=11.0 -fPIC -fvisibility=hidden -Wall -Wextra -fno-exceptions -fPIC -fapplication-extension -MD -MT src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o -MF src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o.d -o src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o -c $SRC_DIR/build/src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm
```

whereas the  successful qt-main build was compiling the same file with cxx:

```
2024-04-10T03:41:37.2727560Z [3702/4296] Building CXX object qtmultimedia/src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_cxx.cxx.o
```

Any ideas on the consequences of removing that unity support for QtMultimedia/is there a better solution you can think of?

I've also tried to clean up extra dependencies for linux for anything not referenced in the CMake variables for the QtMultimedia build, but if it starts to fail again, I'll add the relevant ones back in.

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:86]:
```suggestion
```

**Comment by mmcauliffe** [comment]
Ok this should pass for everything, I added the system dlls that were throwing an error to the missing_dso_whitelist, but let me know if there's a better way to solve that, as well as the question about removing the Unity flags.

Otherwise, feel free to merge and I'll get the cross compilation builds working once the feedstock's made.

**Comment by hmaarrfk** [comment]
can you summarize what you did with osx?

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:20]: is this supposed to be win or unix

**Comment by mmcauliffe** [comment]
I had included this line from qt-main (and qt-charts had it as well): https://github.com/conda-forge/qt-main-feedstock/blob/qt6/recipe/build.sh#L74.  That line was responsible for `unity_0_mm.mm.o` to be built as an OBJCXX object instead of a CXX object:

```
2024-04-10T20:07:59.3806040Z [156/210] Building OBJCXX object src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o
2024-04-10T20:07:59.3909980Z FAILED: src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o 
2024-04-10T20:07:59.4018900Z $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -DGL_SILENCE_DEPRECATION -DQFFmpegMediaPlugin_EXPORTS -DQT_COMPILING_FFMPEG -DQT_CORE_LIB -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_UP_TO=0x050000 -DQT_EXPLICIT_QFILE_CONSTRUCTION_FROM_PATH -DQT_GUI_LIB -DQT_MULTIMEDIA_LIB -DQT_NETWORK_LIB -DQT_NO_AS_CONST -DQT_NO_AS_CONST=1 -DQT_NO_CONTEXTLESS_CONNECT=1 -DQT_NO_DEBUG -DQT_NO_EXCEPTIONS -DQT_NO_FOREACH -DQT_NO_FOREACH=1 -DQT_NO_JAVA_STYLE_ITERATORS -DQT_NO_NARROWING_CONVERSIONS_IN_CONNECT -DQT_NO_QEXCHANGE -DQT_PLUGIN -DQT_USE_QSTRINGBUILDER -DQT_WARN_DEPRECATED_UP_TO=0x070000 -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I$SRC_DIR/build/src/plugins/multimedia/ffmpeg/QFFmpegMediaPlugin_autogen/include -I$SRC_DIR/src/plugins/multimedia/ffmpeg -I$SRC_DIR/build/src/plugins/multimedia/ffmpeg -I$SRC_DIR/build/include -I$SRC_DIR/src/plugins/multimedia/ffmpeg/../darwin -I$SRC_DIR/src/plugins/multimedia/ffmpeg/../darwin/camera -I$SRC_DIR/build/src/multimedia -I$SRC_DIR/build/include/QtMultimedia/6.7.0 -I$SRC_DIR/build/include/QtMultimedia/6.7.0/QtMultimedia -I$SRC_DIR/build/include/QtMultimedia -isystem $PREFIX/include/qt6/QtCore/6.7.0 -isystem $PREFIX/include/qt6/QtCore/6.7.0/QtCore -isystem $PREFIX/include/qt6/QtCore -isystem $PREFIX/include/qt6 -isystem $PREFIX/lib/qt6/mkspecs/macx-clang -isystem $PREFIX/include/qt6/QtGui/6.7.0 -isystem $PREFIX/include/qt6/QtGui/6.7.0/QtGui -isystem $PREFIX/include/qt6/QtGui -isystem $PREFIX/include/qt6/QtNetwork -isystem $PREFIX/include -x objective-c++ -DNDEBUG -O2 -std=c++17 -isysroot /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -mmacosx-version-min=11.0 -fPIC -fvisibility=hidden -Wall -Wextra -fno-exceptions -fPIC -fapplication-extension -MD -MT src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o -MF src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o.d -o src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm.o -c $SRC_DIR/build/src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_mm.mm
```

The successful qt-main build from https://github.com/conda-forge/qt-main-feedstock/pull/256 compiled the same file as a CXX object:

```
2024-04-10T03:41:37.2727560Z [3702/4296] Building CXX object qtmultimedia/src/plugins/multimedia/ffmpeg/CMakeFiles/QFFmpegMediaPlugin.dir/Unity/unity_0_cxx.cxx.o
```

From looking over the QtMultimedia sources, the only reference to Unity is for 3rdparty resonance-audio: https://github.com/qt/qtmultimedia/blob/dev/src/3rdparty/resonance-audio/CMakeLists.txt#L44, which we're not building, so I think it's ok to exclude the Unity flags.

File: recipes/qt-multimedia/meta.yaml
Comment (mmcauliffe) [reply recipes/qt-multimedia/meta.yaml]: Oops, should be fixed now!

**Comment by hmaarrfk** [comment]
understood. well it seems that qt released 6.7.1 today, i'm wondering if this is the window we were looking for to rebuild by splitting out multimedia.

**Comment by mmcauliffe** [comment]
Sure that seems reasonable, do you want me to close the qt-main PR, make a
new one for 6.7.1 and remove QtMulitmedia and its dependencies? And then
once that gets merged, I can update this to target 6.7.1.

**Comment by hmaarrfk** [comment]
Yes. Generally speaking I want to get jschulers active approval too. So we should work to also intercept the bots PR and ping jschulers too

**Comment by mmcauliffe** [comment]
That sounds good, it does look like 6.7.1 still isn't out: https://download.qt.io/official_releases/qt/6.7/, but I'll keep an eye on it.

**Comment by mmcauliffe** [comment]
@hmaarrfk this should be good to go for 6.7.1 now for when we want to start the ball on splitting QtMultimedia off from qt-main.

File: recipes/qt-multimedia/bld.bat
Comment (h-vetinari) [inline recipes/qt-multimedia/bld.bat:1]:
```suggestion
@echo on

```

File: recipes/qt-multimedia/bld.bat
Comment (h-vetinari) [inline recipes/qt-multimedia/bld.bat:16]: I prefer this because it covers all non-zero exit codes and makes it clear that some magic variable is being checked (same for the other places). Also missing newline at the end.
```suggestion
if %ERRORLEVEL% neq 0 exit 1
```

File: recipes/qt-multimedia/build.sh
Comment (h-vetinari) [inline recipes/qt-multimedia/build.sh:1]:
```suggestion
#!/bin/sh
set -ex
```

File: recipes/qt-multimedia/build.sh
Comment (h-vetinari) [inline recipes/qt-multimedia/build.sh:21]: Does that build folder already exist? If not, can we create it to be clear(er)?

File: recipes/qt-multimedia/meta.yaml
Comment (h-vetinari) [inline recipes/qt-multimedia/meta.yaml:47]: Please decide on a selector pattern and use it everywhere.

File: recipes/qt-multimedia/meta.yaml
Comment (h-vetinari) [inline recipes/qt-multimedia/meta.yaml:69]: Is there some discernible order of precedence/priority here? If so, perhaps leave a comment? If not, can we sort them alphabetically?

File: recipes/qt-multimedia/conda_build_config.yaml
Comment (h-vetinari) [inline recipes/qt-multimedia/conda_build_config.yaml:5]: There's a gstreamer migration currently, and qt-main has already moved on, so if you want the build to work right away, you can add the pins from the [migrator](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/migrations/gstreamer124.yaml) here
```suggestion
MACOSX_SDK_VERSION:        # [osx]
  - '12.3'                 # [osx]

gstreamer:
- '1.24'
gst_plugins_base:
- '1.24'
```

**Review by h-vetinari — COMMENTED** [review]
You're in good hands with @hmaarrfk, but just gave this a quick look over.

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:46]:
```suggestion
    - ffmpeg *=lgpl*                     # [build_platform != target_platform]
```

if upstream is discouraging it we can simply remove it from future versions. 

specifying the lgpl build of ffmpeg will help just solidify that you aren't relying on the GPL stuff keeping the license LPGL

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:66]:
```suggestion
    - ffmpeg *=lgpl*
```

if upstream is discouraging it we can simply remove it from future versions. 

specifying the lgpl build of ffmpeg will help just solidify that you aren't relying on the GPL stuff keeping the license LPGL

File: recipes/qt-multimedia/build.sh
Comment (mmcauliffe) [reply recipes/qt-multimedia/build.sh:21]: Hmm, I took this from https://github.com/conda-forge/qt-charts-feedstock/blob/main/recipe/build.sh, do you just want a `mkdir -p build` before the cmake command?

Comment (h-vetinari) [reply recipes/qt-multimedia/build.sh:22]: I think that would be a slight improvement. It's better to know where these folders come from (some projects do have a checked-in `build` folder and this can lead to weird effects), so I'm not super keen on how CMake apparently will create the build folder silently.

Personally I prefer the pattern
```
mkdir build
cd build

cmake [options] \
    ..  # look ma, no -B
```
but I guess it's just a matter of habit.

File: recipes/qt-multimedia/meta.yaml
Comment (h-vetinari) [reply recipes/qt-multimedia/meta.yaml:46]: The pinning is picked up for host, not for build. so this is fine as of https://github.com/conda-forge/staged-recipes/pull/25992/commits/bfc49db2546f366f5bae11e5f2459db644594057

File: recipes/qt-multimedia/meta.yaml
Comment (h-vetinari) [inline recipes/qt-multimedia/meta.yaml:58]: Is there a reason these three libraries only show up in build, but not in host? If yes, can we describe why? If not:
```suggestion
```

File: recipes/qt-multimedia/meta.yaml
Comment (h-vetinari) [inline recipes/qt-multimedia/meta.yaml:68]: Please put comments above the dependency they describe
```suggestion
    # gstreamer plugin not supported on win32: https://bugreports.qt.io/browse/QTBUG-107073
    - gstreamer                          # [unix]
    - gst-plugins-base  {{ gstreamer }}  # [unix]
```

**Review by h-vetinari — COMMENTED** [review]
One last question & a last nit

**Review by h-vetinari — APPROVED** [review]
LGTM!

File: recipes/qt-multimedia/meta.yaml
Comment (hmaarrfk) [inline recipes/qt-multimedia/meta.yaml:121]:
```suggestion
    - mmcauliffe
    - conda-forge/qt-main
```

**Review by hmaarrfk — COMMENTED** [review]
I'll press merge when it is green again!! Thanks for the patience

**Comment by hmaarrfk** [comment]
I confirm on behalf of @conda-forge/qt-main that I wish to maintain this recipe

**Comment by hmaarrfk** [comment]
fingers crossed.

**Pull request merged**
